### DISCUSSION TITLE
I really appreciate the design of this project

### DISCUSSION THREAD

**Discussion by Prox501** [pfusik/cito#76] [General]
I came across this project and it offers majority of the features that I was looking for out of the box. Granted there are some things I had to change in the implementation of Cito programming language (which effectively make it a new programming language.) I removed majority of the language targets that Cito compiles to and have it compiled only for C programming language.

Basically the goals are as followed:

1. Change unsigned integer to use the full extent of 32 bit unsigned integer rather than 31 bit as well as supporting unsigned 64 bit integer and various SIMD types.
2. Implement Getter/Setter properties (Basically adding getter/setter methods into virtual table and implementations)
3. Allow distinction between Private, and Public visibility that is arranged as this:
    - public visibility = include/*
    - private visibility = include/private/*
    - all visibilites = src/*
4. Allow external function declarations to be defined to allow Cito to compile.
5. Break up Cito files into it's own individual files rather than generating into a singular output file.
6. Enable parameter overloading in constructors/functions by appending number at the end of the function name such as BaseObject_new1(), BaseObject_new2(int a), and so forth.
7. Generates JSON file for binding C project (effectively allowing **all** programming language to use the resulting work.)
8. Language Server Protocol support (That I could use with Kate editor.)
9. Automatically generates Virtual Dispatch functions to hide Virtual Table pointer going forward.
10. Get rid of GLibC implementations by using my personally written STL library for C Language. Basically supporting Musl C Library rather than just GLibc.

Basically I am making Cito something closer to a full blown programming language at some point and I really appreciates that the code structure for this project is really clean and easily extensible. 👍🏻

**Comment by pfusik** [comment]
Thanks for sharing this! So you basically forked cito, restricted it to C output only and added the listed features? Where can we find your project?

**Reply by Prox501** [reply]
Yep, that the general idea for Cito, making it compile to C only and have the listed features. I figured that if it have C API and JSON file for binding the library, it would mostly address the same problem that Cito aims to solve.

Likely on Codeberg, I plan on uploading there sometime later in the weekend. And it's definitely far from ready, but this project make it a lot more doable.

Here my other project and I will have the language uploaded on the same account: https://codeberg.org/TylerLinuxDev/STL-For-C

I've been looking on the internet for solutions to address the headache of having to write OOP C programming by hands and Cito basically gave me an opportunity to solve this and bringing back a little bit C# into C programming. 😄

**Reply by Prox501** [reply]
I probably would call this forked language, "Clight", which seems not to be taken yet. Essentially a forked Cito langauge for C Object Oriented Programming Langauge transpiler with support for easier binding process via JSON.

**Reply by Prox501** [reply]
As promised, still work in progress: https://codeberg.org/TylerLinuxDev/Clight

**Comment by pfusik** [comment]
There are four commits at https://codeberg.org/TylerLinuxDev/Clight/commits/branch/master since the fork, and 600+ here. Basically, these are two very different projects now. I would like to comment on your goals:

1. 32-bit `uint` and 64-bit `ulong` are tracked by #61.
2. That would be useful in C# and JavaScript outputs. I'll consider it.
3. I see no reason why `private` members should be in a different file than the `.c` implementation. Ability to specify a different directory for the header file might be useful.
4. My priority is having a language for portable implementation of low-level components. Interfacing with the target language (using `native`) is an option, but of lower priority. Since Clight only targets C, it's a different story.
5. It's not clear to me how to break and why?
6. https://github.com/fusionlanguage/fut/releases/tag/cito-2.1.1 introduces object literals, which should cover typical usecases for overloaded constructors.
7. What JSON file?
8. I have this in plans, but first I'd like to continue work on #64. Currently it uses the parser in Fusion transpiled to JavaScript to report syntax errors.
9. Calling `virtual` methods from C is one problem. Overriding Fusion methods from C is a bigger one. The current solution is not very elegant.
10. Did you mean glib as opposed to glibc? There was a short discussion (#91), but for now, glib works fine for me.
